File: 139.单词拆分.cpp
// 可以用无限次
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        if(wordDict.size() == 0)
            return false;
        int n = s.length();
        int* dp = new int[n + 1]();
        dp[0] = 1;
        // dict中的单词没有使用限制，因此是一个完全背包问题
        for(int i = 1; i <= n; ++i){
            for(string word: wordDict){
                int len = word.length();
                if(len <= i && word == s.substr(i - len, len) && (dp[i] || dp[i - len]))
                    dp[i] = 1;
            }
        }
        return dp[n];
    }
};

File: 动态规划dp/139.单词拆分.js
/**
 * @param {string} s
 * @param {string[]} wordDict
 * @return {boolean}
 */
var wordBreak = function (s, wordDict) {
    let n = s.length;
    // dp[i] 表示字符串 s 的前 i 个字符能否拆分成  wordDict 中的单词。
    let dp = new Array(n + 1).fill(false);
    dp[0] = true;
    for (let i = 1; i <= n; ++i) {
        for (let word of wordDict) {
            let len = word.length;
            if (len <= i && word === s.slice(i - len, i) && dp[i - len]) {
                dp[i] = true;
                break;
            }
        }
    }
    return dp[n];
};

var wordBreak = function (s, wordDict) {
    let n = s.length;
    let dp = new Array(n + 1).fill(false);
    dp[0] = true;
    for (let i = 1; i <= n; ++i) {
        for (let j = 0; j < i; ++j) {
            if (wordDict.indexOf(s.slice(j, i)) !== -1 && dp[j]) {
                dp[i] = true;
                break;
            }
        }
    }
    return dp[n];
};